### DISCUSSION TITLE
Main vs master-using git push and behind the scene behaviour of github if not renaming local branch to main

### DISCUSSION THREAD

**Discussion by adi-garg** [community/community#47878] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

General

### Body

Hello Community,
Just started learning Git/Github.
I notice that:

(As of 2023 where github calls the first branch as _main_ instead of _master_)

The command `git push -u origin master` implies use **master** of _local_ and push into **master** of _github_.

Similarly,`git push -u origin main` implies use **main** of local and push into **main** of _github_.

**Scenario where local branch name is not updated to _main_:**

As a first intuition, with _main_ being the default branch on github,`git push -u origin master` command should ideally end up creating 2 branches on github-one called _main_ (as default behaviour which exists as first branch) and one NEW called _master_(due to the command)?

**What happens as observed practically- GitHub smartly renames its newer main branch to older master  branch the moment above command is issued.**

Secondly,if one were to issue `git push -u origin main` ,it complains(error) that there is such local branch called _main_:

_error: src refspec main does not match any_

Wanting to revalidate if the above understanding is correct. Was surprised that there is not much discussion online on the `git push -u origin main` command inference(local vs remote branch name) and on behind the scenes behaviour of github(reverting name _main_ to _master_).

Kind regards,
Aditya Garg

**Comment by airtower-luna** [comment]
This is described in detail in the [`git push` manual page](https://git-scm.com/docs/git-push). In your examples `main` or `master` is the *refspec*, which tells Git which local ref (tag or branch) to push to which remote ref. Using the branch name is the most common and simplest option: The branch is pushed to the remote branch of the same name.

But the refspec can be a lot more complex. In your case you probably want something like `master:main` to push the local ref `master` to the remote ref `main`.

That said, the simplest option is probably to rename your local branch, too: `git branch -m master main`. :slightly_smiling_face:

**Reply by adi-garg** [reply]
Thank you for validating the understanding of syntax. Do you also notice the behind the scene behaviour of Github-reverting _main_ to _master_ on issuing `git push origin master` (instead of creating new branch called _master_ **in addition to** existing _main_)?

**Reply by airtower-luna** [reply]
The only case where I know something similar is when you create a new empty repository: The first branch you push becomes the default branch.

**Reply by adi-garg** [reply]
Thank you, I got this validated practically and it can be viewed in terms of 2 scenarios:
<NOT Cloning locally below but having push a local branch from local repo to github's newly created repo--Cloning would not cause any issue as the branch would already have the name _main_ and same clones locally>

Scenario where **local branch name is not updated to main:**

1.If the github repository has had **atleast one commit**(for example initialised with readme on github):

Then, `git push -u origin master `would cause a new branch called **master** to be created on github,in additional to **main** branch on github.

2.If the github repository has **0 commits**:

It should be noted that there is no branch(**main**) existent yet on Github.

`git push -u origin master` command causes a new branch called **master** to be created.

Infact,any rename done locally,example: `git branch -m master trunk` followed by `git push -u origin trunk` actually causes github to use that first branch ie **trunk** here.
